packages/dashboard-core-plugins/src/panels/PanelContextMenu.tsx

```diff
@@ -102,11 +123,26 @@ class PanelContextMenu extends PureComponent<
   }
 
   render(): ReactElement {
-    const { additionalActions, glContainer } = this.props;
+    const { additionalActions, workspace, glContainer } = this.props;
 
     const contextActions: (ContextAction | (() => ContextAction))[] = [
       ...additionalActions,
     ];
+    const stackId = LayoutUtils.getStackForConfig(
+      glContainer.layoutManager.root,
+      glContainer.getConfig()
+    )?.config.id;
+    const hasPanelToReopen = workspace.data.closed?.some(
+      panel => (panel as ReactComponentConfig).parentStackId === stackId
+    );
+
+    contextActions.push(() => ({
+      title: 'Re-open closed panel',
+      order: 10,
+      group: ContextActions.groups.reopen,
+      action: this.handleReopenLast,
+      disabled: !hasPanelToReopen,
+    }));
```

Comment:
The check for `hasPanelToReopen` is fairly expensive - you should plumb this up like the disabled check below for `this.canCloseTabsRight()`, for example.

Right now you're running the check (which does `LayoutUtils.getStackForConfig`, `workspace.data.closed?.some(...)`, both expensive operations depending on how many components in the layout/workspace data) on every `render`. Try setting a breakpoint in the panel and see how often it re-renders (moving panels, updating table filters, maybe some other things?).

You only need to check when the context menu appears, which you can do by just passing in a function action and checking in there.